--- a/plugins/password/drivers/ldap.php.orig	2014-09-29 14:56:50.000000000 +0800
+++ b/plugins/password/drivers/ldap.php	2014-11-08 13:50:31.000000000 +0800
@@ -283,6 +283,58 @@
             }
             break;
 
+        case 'ssha512':
+            $rcmail = rcmail::get_instance();
+            $dovecotpw = $rcmail->config->get('password_dovecotpw');
+            $method = $rcmail->config->get('password_dovecotpw_method');
+
+            // use common temp dir
+            $tmp_dir = $rcmail->config->get('temp_dir');
+            $tmpfile = tempnam($tmp_dir, 'roundcube-');
+
+            $pipe = popen("$dovecotpw -s '$method' -p '$password_clear'> '$tmpfile'", "w");
+            if (!$pipe) {
+                unlink($tmpfile);
+                return PASSWORD_CRYPT_ERROR;
+            }
+            else {
+                pclose($pipe);
+                $crypted_password = trim(file_get_contents($tmpfile), "\n");
+                if (!preg_match('/^\{' . $method . '\}/', $crypted_password)) {
+                    return PASSWORD_CRYPT_ERROR;
+                }
+                if (!$rcmail->config->get('password_dovecotpw_with_method'))
+                    $crypted_password = trim(str_replace('{' . $method . '}', '', $crypted_password));
+                unlink($tmpfile);
+                break;
+            }
+        case 'blf-crypt':
+            $rcmail = rcmail::get_instance();
+            $dovecotpw = $rcmail->config->get('password_dovecotpw');
+            $method = $rcmail->config->get('password_dovecotpw_method');
+
+            // use common temp dir
+            $tmp_dir = $rcmail->config->get('temp_dir');
+            $tmpfile = tempnam($tmp_dir, 'roundcube-');
+
+            $pipe = popen("$dovecotpw -s '$method' -p '$password_clear'> '$tmpfile'", "w");
+            if (!$pipe) {
+                unlink($tmpfile);
+                return PASSWORD_CRYPT_ERROR;
+            }
+            else {
+                pclose($pipe);
+                $crypted_password = trim(file_get_contents($tmpfile), "\n");
+                if (!preg_match('/^\{' . $method . '\}/', $crypted_password)) {
+                    return PASSWORD_CRYPT_ERROR;
+                }
+                if (!$rcmail->config->get('password_dovecotpw_with_method'))
+                    $crypted_password = trim(str_replace('{' . $method . '}', '', $crypted_password));
+                if ($method == 'BLF-CRYPT')
+                    $crypted_password = trim(str_replace('{' . $method . '}', '{CRYPT}', $crypted_password));
+                unlink($tmpfile);
+                break;
+            }
         case 'smd5':
             mt_srand((double) microtime() * 1000000);
             $salt = substr(pack('h*', md5(mt_rand())), 0, 8);
