--- a/plugins/password/drivers/sql.php	2014-09-29 14:56:50.000000000 +0800
+++ b/plugins/password/drivers/sql.php	2014-11-08 10:13:00.000000000 +0800
@@ -95,21 +95,25 @@
             $tmp_dir = $rcmail->config->get('temp_dir');
             $tmpfile = tempnam($tmp_dir, 'roundcube-');
 
-            $pipe = popen("$dovecotpw -s '$method' > '$tmpfile'", "w");
+            $pipe = popen("$dovecotpw -s '$method' -p '$passwd' > '$tmpfile'", "w");
             if (!$pipe) {
                 unlink($tmpfile);
                 return PASSWORD_CRYPT_ERROR;
             }
             else {
-                fwrite($pipe, $passwd . "\n", 1+strlen($passwd)); usleep(1000);
-                fwrite($pipe, $passwd . "\n", 1+strlen($passwd));
+                //fwrite($pipe, $passwd . "\n", 1+strlen($passwd)); usleep(1000);
+                //fwrite($pipe, $passwd . "\n", 1+strlen($passwd));
                 pclose($pipe);
                 $newpass = trim(file_get_contents($tmpfile), "\n");
+
                 if (!preg_match('/^\{' . $method . '\}/', $newpass)) {
                     return PASSWORD_CRYPT_ERROR;
                 }
                 if (!$rcmail->config->get('password_dovecotpw_with_method'))
                     $newpass = trim(str_replace('{' . $method . '}', '', $newpass));
+                if ($method == 'BLF-CRYPT') {
+                    $newpass = trim(str_replace('{' . $method . '}', '{CRYPT}', $newpass));
+                }
                 unlink($tmpfile);
             }
             $sql = str_replace('%D', $db->quote($newpass), $sql);
