--- a/plugins/password/drivers/sql.php	2014-09-29 14:56:50.000000000 +0800
+++ b/plugins/password/drivers/sql.php	2014-11-08 10:39:41.000000000 +0800
@@ -81,7 +81,7 @@
                 $salt .= $seedchars[rand(0, 63)];
             }

-            $sql = str_replace('%c',  $db->quote(crypt($passwd, $salt_hashindicator ? $salt_hashindicator .$salt.'$' : $salt)), $sql);
+            $sql = str_replace('%c', str_replace("'", "\'", $db->quote(crypt($passwd, $salt_hashindicator ? $salt_hashindicator .$salt.'$' : $salt))), $sql);
         }

         // dovecotpw
@@ -95,14 +95,12 @@
             $tmp_dir = $rcmail->config->get('temp_dir');
             $tmpfile = tempnam($tmp_dir, 'roundcube-');

-            $pipe = popen("$dovecotpw -s '$method' > '$tmpfile'", "w");
+            $pipe = popen("$dovecotpw -s '$method' -p '$passwd' > $tmpfile", "w");
             if (!$pipe) {
                 unlink($tmpfile);
                 return PASSWORD_CRYPT_ERROR;
             }
             else {
-                fwrite($pipe, $passwd . "\n", 1+strlen($passwd)); usleep(1000);
-                fwrite($pipe, $passwd . "\n", 1+strlen($passwd));
                 pclose($pipe);
                 $newpass = trim(file_get_contents($tmpfile), "\n");
                 if (!preg_match('/^\{' . $method . '\}/', $newpass)) {
@@ -110,9 +108,11 @@
                 }
                 if (!$rcmail->config->get('password_dovecotpw_with_method'))
                     $newpass = trim(str_replace('{' . $method . '}', '', $newpass));
+                if ($method == 'BLF-CRYPT')
+                    $newpass = trim(str_replace('{' . $method . '}', '{CRYPT}', $newpass));
                 unlink($tmpfile);
             }
-            $sql = str_replace('%D', $db->quote($newpass), $sql);
+            $sql = str_replace('%D', str_replace("'", "\'", $db->quote($newpass)), $sql);
         }

         // hashed passwords
@@ -178,7 +178,7 @@
         // at least we should always have the local part
         $sql = str_replace('%l', $db->quote($local_part, 'text'), $sql);
         $sql = str_replace('%d', $db->quote($domain_part, 'text'), $sql);
-        $sql = str_replace('%u', $db->quote($username, 'text'), $sql);
+        $sql = str_replace('%u', str_replace("'", "\'", $db->quote($username, 'text')), $sql);
         $sql = str_replace('%h', $db->quote($host, 'text'), $sql);
 
         $res = $db->query($sql, $sql_vars);
